src/N3Index.js

```diff
@@ -0,0 +1,604 @@
+            if (key2) (key2 in index2) && count++;
+            // Otherwise, count all quads
+            else count += Object.keys(index2).length;
+          }
+        }
+      }
+    }
+    return count;
+  }
+
+  // ### `_findInIndex` finds a set of quads in a three-layered index.
+  // The index base is `index0` and the keys at each level are `key0`, `key1`, and `key2`.
+  // Any of these keys can be undefined, which is interpreted as a wildcard.
+  // `name0`, `name1`, and `name2` are the names of the keys at each level,
+  // used when reconstructing the resulting quad
+  // (for instance: _subject_, _predicate_, and _object_).
+  // Finally, `graph` will be the graph of the created quads.
+  // If `callback` is given, each result is passed through it
+  // and iteration halts when it returns truthy for any quad.
+  // If instead `array` is given, each result is added to the array.
+  _findInIndex(index0, key0, key1, key2, name0, name1, name2, graph, callback, array) {
+    let tmp, index1, index2;
+    // TODO: See if Object.keys(this._ids) should be optimized. this._ids only changes
+    // in addQuad and removeQuad. What are the performance implications if a Map is
+    // used instead of JS objects? Will difference between get, set, remove outweigh
+    // perf improvement of Map#size compared to Object.keys(obj)#length?
+    // Depending on the number of variables, keys or reverse index are faster
+    const varCount = !key0 + !key1 + !key2,
+        entityKeys = varCount > 1 ? Object.keys(this._ids) : this._entities;
+
+    // If a key is specified, use only that part of index 0.
+    if (key0) (tmp = index0, index0 = {})[key0] = tmp[key0];
+    for (const value0 in index0) {
+      // TODO: Move entity0 into the next if.
+      const entity0 = entityKeys[value0];
+
+      if (index1 = index0[value0]) {
+        // If a key is specified, use only that part of index 1.
+        if (key1) (tmp = index1, index1 = {})[key1] = tmp[key1];
+        for (const value1 in index1) {
+          // TODO: Move entity1 into the next if.
+          const entity1 = entityKeys[value1];
+
+          if (index2 = index1[value1]) {
+            // If a key is specified, use only that part of index 2, if it exists.
+            const values = key2 ? (key2 in index2 ? [key2] : []) : Object.keys(index2);
+            // Create quads for all items found in index 2.
+            for (let l = 0; l < values.length; l++) {
+              const parts = { subject: null, predicate: null, object: null };
+              // TODO: Only call termFromId once for each term if it can be done efficiently.
+              parts[name0] = termFromId(entity0, this._factory);
+              parts[name1] = termFromId(entity1, this._factory);
+              parts[name2] = termFromId(entityKeys[values[l]], this._factory);
+
+              // TODO: Convert this to a generator function and yield each quad, but yield
+              // quad tuples if that is more efficient.
```

Comment:
I'm thinking it would be good to change this `_findInIndex` method to a generator function and yield each quad or a quad tuple (an array of subject, predicate, object, and graph). Then the code on lines 524-527 can be moved into N3Store to "handle" the yielded quad. This way the N3Index doesn't need to know what needs to be done with the quad besides yielding it.